### PR TITLE
gh-96192: fix os.ismount() to use a path that is str or bytes

### PR DESCRIPTION
**Pull request by calestyo**
Co-authored-by: Eryk Sun <eryksun@gmail.com>
Signed-off-by: Christoph Anton Mitterer <mail@christoph.anton.mitterer.name>

* Issue: gh-96192

### CODE REVIEW

**Comment by orsenthil** [comment]
Hi @calestyo , Would you like to add a test for this?

**Comment by calestyo** [comment]
@orsenthil Well I could at least try, but two issues here:

- I think it would be handy to have a test that does this for **any** function that is ought to accept the PathLike-interface, but that in turn doesn't seem to be possibly generically.
- The test get a bit ugly, since one cannot create `os.DirEntry` objects, even if I'd just made one for `os.ismount()`. Right now these do:
https://github.com/python/cpython/blob/e34c82abeb7ace09e6b5d116585c47cc372996c1/Lib/test/test_posixpath.py#L178-L180

but one cannot yield `/` as `os.DirEntry` via `scandir()`.

**Comment by orsenthil** [comment]
Without this patch, I would expect this

> self.assertIs(posixpath.ismount(b"/"), True) 

to fail or behavior differently than with this patch. 

What might I be missing?

**Comment by calestyo** [comment]
> Without this patch, I would expect this
> 
> > self.assertIs(posixpath.ismount(b"/"), True)
> 
> to fail or behavior differently than with this patch.

That I don't understand… as far as I understand it, it should have worked previously with both `str` and `bytes` but not when e.g. using it with a `os.DirEntry`.

File: Misc/NEWS.d/next/Library/2022-08-23-03-13-18.gh-issue-96192.TJywOF.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fixed the :func:`os.ismount()` to use a path that is str or bytes in all cases.
```
Comment (JelleZijlstra) [inline Misc/NEWS.d/next/Library/2022-08-23-03-13-18.gh-issue-96192.TJywOF.rst]:
```suggestion
Fix handling of ``bytes`` :term:`path-like objects <path-like object>` in :func:`os.ismount()`.
```

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
Could you add a test that passes a bytes path-like to `ismount()`?

File: Misc/NEWS.d/next/Library/2022-08-23-03-13-18.gh-issue-96192.TJywOF.rst (hunk 1 above)
Comment (calestyo) [reply Misc/NEWS.d/next/Library/2022-08-23-03-13-18.gh-issue-96192.TJywOF.rst]: Updated in the current PR, added you as Co-Author.

**Comment by calestyo** [comment]
> Could you add a test that passes a bytes path-like to ismount()?

@orsenthil  @JelleZijlstra … I've added a test using `Path`. Does that seem ok for you?

**Comment by calestyo** [comment]
I have made the requested changes; please review again

**Comment by JelleZijlstra** [comment]
The tests are failing.

**Comment by calestyo** [comment]
Uagh... seems `Path` itsellf claims it would accept a PathLike object, e.g. `pathlib.PurePath(*pathsegments)` says:
> Each element of pathsegments can be either a string representing a path segment, an object implementing the os.PathLike interface which returns a string, or another path object:

but it doesn't really, but only such that use strings. Isn't that yet another bug?

Anyway... I don't know how to do a proper test then. Is there any other PathLike object type in the library?

The only I'd know is `os.DirEntry` which one cannot directly create. I could use `os.scandir(b"/")__next__()` but there's the result could be a mointpoint or not.

Any ideas?

**Comment by calestyo** [comment]
Oh and is a test here really worth it? The fix is trivial, should someone ever remove it, the check could also just be removed by accident as well.

As I've already said [above](https://github.com/python/cpython/pull/96194#issuecomment-1226635284), to which noone replied,... it would IMO only make sense, if **all** these functions could be tested for whether they accept the right types, at once.

**Comment by JelleZijlstra** [comment]
> Uagh... seems `Path` itsellf claims it would accept a PathLike object, e.g. `pathlib.PurePath(*pathsegments)` says:
> 
> > Each element of pathsegments can be either a string representing a path segment, an object implementing the os.PathLike interface which returns a string, or another path object:
> 
> but it doesn't really, but only such that use strings. Isn't that yet another bug?

The problem in your test is that `Path` only supports str paths, not bytes paths. This is well-established behavior and not a bug.

> Anyway... I don't know how to do a proper test then. Is there any other PathLike object type in the library?

You can use `test.support.os_helper.FakePath`.

> Oh and is a test here really worth it?

Yes, every bugfix needs an associated test.

**Comment by calestyo** [comment]
> You can use test.support.os_helper.FakePath.

thx... still fails on one platform though... no idea why

**Comment by JelleZijlstra** [comment]
Looks spurious, I retried it.

**Review by JelleZijlstra — APPROVED** [review]

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**